src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs

```diff
@@ -152,7 +147,111 @@ public IEnumerable<IFeatureInfo> GetDependentFeatures(string featureId)
             })).Value;
         }
 
-        private IEnumerable<IFeatureInfo> GetFeatureDependencies(
+        private void EnsureInitialized()
+        {
+            if (_isInitialized)
+            {
+                return;
+            }
+
+            lock (_synLock)
+            {
+                if (_isInitialized)
+                {
+                    return;
+                }
+
+                var applicationContext = _serviceProvider.GetRequiredService<IApplicationContext>();
+                var typeFeatureProvider = _serviceProvider.GetRequiredService<ITypeFeatureProvider>();
+                var featuresProvider = _serviceProvider.GetRequiredService<IFeaturesProvider>();
+
+                var extensionDependencyStrategies = _serviceProvider.GetServices<IExtensionDependencyStrategy>();
+                var extensionPriorityStrategies = _serviceProvider.GetServices<IExtensionPriorityStrategy>();
+
+                var modules = applicationContext.Application.Modules;
+                var loadedExtensions = new ConcurrentDictionary<string, ExtensionEntry>();
+
+                // Load all extensions in parallel
+                Parallel.ForEach(modules, (module, cancellationToken) =>
+                {
+                    if (!module.ModuleInfo.Exists)
+                    {
+                        return;
+                    }
+
+                    var manifestInfo = new ManifestInfo(module.ModuleInfo);
+                    var extensionInfo = new ExtensionInfo(module.SubPath, manifestInfo, (mi, ei) =>
+                    {
+                        return featuresProvider.GetFeatures(ei, mi);
+                    });
+
+                    var entry = new ExtensionEntry
+                    {
+                        ExtensionInfo = extensionInfo,
+                        Assembly = module.Assembly,
+                        ExportedTypes = module.Assembly.ExportedTypes
+                    };
+
+                    loadedExtensions.TryAdd(module.Name, entry);
+                });
+
+                // Get all types from all extension and add them to the type feature provider.
+                foreach (var loadedExtension in loadedExtensions)
+                {
+                    var extension = loadedExtension.Value;
+
+                    foreach (var exportedType in extension.ExportedTypes.Where(IsComponentType))
```

Comment:
Better to save the components once so it's not evaluated on each iteration. 
```suggestion
                var componentTypes = extension.ExportedTypes.Where(IsComponentType).ToArray();
                
                // Get all types from all extension and add them to the type feature provider.
                foreach (var componentType in componentTypes)
                {
                    var extension = componentType.Value;

                    foreach (var exportedType in componentTypes)
```

Reply:
This is not necessary as it's iterated only once.

Reply:
Its a loop inside another loop, meaning it could iterate more than once. Otherwise, why the nested loop?